### PR TITLE
[ATL-959] Remove settings icon from sidebar

### LINKED ISSUE #15: Move settings access to File > Preferences
**Is your feature request related to a problem? Please describe.**
A pretty gear icon is shown prominently in the side bar:
Clipboard02
When this is clicked, the very complex "Preferences" tab opens. This exposes options that will be of no use to the target user. The settings available via **File > Preferences** are the only thing the target user needs. It is good to give more advanced users access to advanced configuration options (as is done in the Java IDE's  **File > Preferences > More preferences can be edited directly in the file...** but the current UI funnels all users equally to those advanced preferences.

**Describe the solution you'd like**
Make the gear icon in the side bar open the **File > Preferences** dialog. Provide some form of access (even if it's only mentioning editing the settings.json file directly) to the advanced configuration from **File > Preferences**.

**Describe alternatives you've considered**
Remove the gear icon from the side bar. Preferences are accessed only via **File > Preferences**.

**Environment**

OS: Windows 10
Version: 2.0.0-beta.2-snapshot.298cc11
Date: 2021-02-15T14:16:52.603Z
CLI Version: 0.16.0 alpha [c977a238]

### PR DESCRIPTION
**Pull request by fstasi**
A gear icon is shown prominently in the side bar

This PR removes the gear icon from the sidebar

[Jira Task](https://arduino.atlassian.net/browse/ATL-959)
Github Issue: fixes #15

### CODE REVIEW

File: arduino-ide-extension/src/browser/contributions/settings.ts [hunk 1]
```diff
@@ -56,5 +66,11 @@ export namespace Settings {
             label: 'Open Preferences...',
             category: 'Arduino'
         }
+
+        export const CLOSE: Command = {
+            id: 'arduino-settings-close',
+            label: 'Close Preferences...',
+            category: 'Arduino'
+        }
```
Comment (fstasi) [inline arduino-ide-extension/src/browser/contributions/settings.ts:74]: I don't think this approach is really necessary as the Dialog provides a Promise than, when resolved (usually with `close()` closes the UI.

File: arduino-ide-extension/src/browser/settings.tsx [hunk 2]
```diff
@@ -376,9 +378,18 @@ export class SettingsComponent extends React.Component<SettingsComponent.Props,
                     onChange={this.additionalUrlsDidChange} />
                 <i className='fa fa-window-restore theia-button shrink' onClick={this.editAdditionalUrlDidClick} />
             </div>
+
+            <div className='flex-line'>
+                <button className='theia-button shrink' onClick={this.openExtendedSettings}>Extended Settings</button>
+            </div>
```
Comment (fstasi) [inline arduino-ide-extension/src/browser/settings.tsx:384]: @kittaakos This is just a draft button in the Arduino settings panel. When clicked it opens the "theia" settings in a tab, and - supposedly - closes the current Arduino settings panel.

I don't really like the way I close this dialog, as I don't think I need an extra command, as the `SettingsComponent` shouldn't be aware it's called inside a dialog.

In my opinion a better approach should be sending a message to the parent (Widget) stating that the extended settings are being opened. The Widget should communicate somehow to the Dialog to call the `.close()`.

File: arduino-ide-extension/src/browser/settings.tsx [hunk 3]
```diff
@@ -376,9 +378,18 @@ export class SettingsComponent extends React.Component<SettingsComponent.Props,
...
         </div>;
     }
 
+    protected openExtendedSettings = (e: React.MouseEvent<HTMLElement>) => {
+      this.props.commands.executeCommand(CommonCommands.OPEN_PREFERENCES.id);
+      this.props.commands.executeCommand('arduino-settings-close');
```
Comment (fstasi) [inline arduino-ide-extension/src/browser/settings.tsx:390]: Even if I defined `CommonCommands.CLOSE_PREFERENCES` I'm not even able to use it here, as it causes a loop in the dependencies.

If we go the "commands path" we should move these preference commands to another file, but my understanding is we don't do that usually.

File: arduino-ide-extension/src/browser/contributions/settings.ts (hunk 1 above)
Comment (kittaakos) [reply arduino-ide-extension/src/browser/contributions/settings.ts:73]: I agree. We do not need a command for it. Especially, a command that is available from the _Commnad Palette_. Note: if you add a `label` to the command, users can execute it from the _Commad Palette_.

File: arduino-ide-extension/src/browser/settings.tsx [hunk 4]
```diff
@@ -13,6 +13,8 @@ import { FileService } from '@theia/filesystem/lib/browser/file-service';
 import { ThemeService } from '@theia/core/lib/browser/theming';
 import { MaybePromise } from '@theia/core/lib/common/types';
 import { WindowService } from '@theia/core/lib/browser/window/window-service';
+import { CommandRegistry } from "@theia/core";
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/settings.tsx:16]: Please use single quotes. I have to add linter for it. Thanks!

File: arduino-ide-extension/src/browser/settings.tsx (hunk 2 above)
Comment (kittaakos) [reply arduino-ide-extension/src/browser/settings.tsx:384]: > as I don't think I need an extra command

I agree. Let's do it without a command.

File: arduino-ide-extension/src/browser/settings.tsx [hunk 5]
```diff
@@ -681,12 +693,17 @@ export class SettingsWidget extends ReactWidget {
     @inject(WindowService)
     protected readonly windowService: WindowService;
 
+    @inject(CommandRegistry)
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/settings.tsx:696]: FYI: There is the `CommandService` API, which's sufficient for this case, as you want to only execute a command. No need to change, just wanted to share this with you.

File: arduino-ide-extension/src/browser/settings.tsx (hunk 4 above)
Comment (fstasi) [reply arduino-ide-extension/src/browser/settings.tsx:16]: nice, we don't use prettier or editorconfig, correct? do we plan to use it in the future?

File: arduino-ide-extension/src/browser/contributions/settings.ts (hunk 1 above)
Comment (fstasi) [reply arduino-ide-extension/src/browser/contributions/settings.ts:73]: thanks for the tip about the `label`!

File: arduino-ide-extension/src/browser/settings.tsx (hunk 5 above)
Comment (fstasi) [reply arduino-ide-extension/src/browser/settings.tsx:696]: 👍 I'm gonna check the difference between the `CommandService` and the `CommandRegistry`

**Comment by kittaakos** [comment]
I have two questions regarding the UX, I guess you have already discussed it with the designers:
 - What should happen when the user makes a few changes in the dialog and opens the advanced settings: do we save the changes before opening the advanced settings editor and closing the dialog or not?
 - If we have to persist the changes before opening the advanced settings, what should happen when the user set invalid settings. Does the dialog show the validation errors before opening the advanced settings and closing itself?
 - What should happen when the advanced editor is already open and the user opens it?

Thanks!

FYI: if the user changes anything in the advanced settings editor, it is very likely that the change goes to the workspace settings. There will be a new `.theia/settings.json` file in the sketch folder.

**Comment by ubidefeo** [comment]
>* What should happen when the user makes a few changes in the dialog and opens the advanced settings: do we save the changes before opening the advanced settings editor and closing the dialog or not?

any change to the Arduino preferences before the advanced preferences should be saved

>* If we have to persist the changes before opening the advanced settings, what should happen when the user set invalid settings. Does the dialog show the validation errors before opening the advanced settings and closing itself?

which setting would be invalid? for instance characters in the scaling percentage?

>* What should happen when the advanced editor is already open and the user opens it?

as far as I understand that advanced settings tab is just a GUI rendering of a setting file, and by trying to open it twice from the gear icon it will just receive focus, not open into another tab

> FYI: if the user changes anything in the advanced settings editor, it is very likely that the change goes to the workspace settings. There will be a new .theia/settings.json file in the sketch folder.

excellent point. I have to say that while for basic users that `.theia` folder would never even show up since it's hidden and I personally don't have a problem with advanced users having it there, but it breaks the sketch paradigm by introducing hidden files.
I think we should just bite the bullet and accept that you can have such things. @alranel was the one proposing to remove it in favour of a temporary path not to break the Sketch format, so let's see what his point is about this.

Thank you for now 👍🏼

**Comment by fstasi** [comment]
> > * What should happen when the user makes a few changes in the dialog and opens the advanced settings: do we save the changes before opening the advanced settings editor and closing the dialog or not?
> 
> any change to the Arduino preferences before the advanced preferences should be saved
> 
> > * If we have to persist the changes before opening the advanced settings, what should happen when the user set invalid settings. Does the dialog show the validation errors before opening the advanced settings and closing itself?
> 
> which setting would be invalid? for instance characters in the scaling percentage?
> 
> > * What should happen when the advanced editor is already open and the user opens it?
> 
> as far as I understand that advanced settings tab is just a GUI rendering of a setting file, and by trying to open it twice from the gear icon it will just receive focus, not open into another tab
> 

We had a chat @giannicolab and I about the "open extended preferences from standard preferences" about how user friendly this flow is.
Having the "extended preferences" button in the "quick" preferences opens up a number number of unclear scenarios (cleverly pointed out by Akos) that are somehow the direct consequence of a bad design choice in the first place.

@giannicolab is suggesting to have 2 menu items, such as "quick preferences" and "extended preferences", rather than having a button in the quick preferences that closes a modal and radically change how the user should interact with the settings. At the moment I think his approach is the best for the user, removing noise from the quick settings (no more extended button) and having to interaction flows which are distinct and separated.

@ubidefeo what is your opinion on this?

**Comment by kittaakos** [comment]
> which setting would be invalid? for instance characters in the scaling percentage?

Yes. Or a completely invalid path pointing to the sketchbook.

> any change to the Arduino preferences before the advanced preferences should be saved

@fstasi, if the changes have to be saved, we need `accept()` here:
https://github.com/arduino/arduino-ide/blob/0ac603045d2c42bb21abb9eb06db1b0cfb263d3d/arduino-ide-extension/src/browser/settings.tsx#L762

**Review by ubidefeo — APPROVED** [review]
LGTM
I tested that the advanced settings helper came up with the shortcut.
we're good

**Pull request merged**